src/flint/flint_base/flint_base.pyx

```diff
@@ -114,16 +119,117 @@ cdef class flint_poly(flint_elem):
                 v = - fac[0]
                 roots.append((v, m))
         return roots
-    
+
     def complex_roots(self):
         raise AttributeError("Complex roots are not supported for this polynomial")
 
 
+cdef class flint_mpoly_context(flint_elem):
+    """
+    Base class for multivariate ring contexts
+    """
+
+    _ctx_cache = None
+
+    def __cinit__(self):
+        self._init = False
+
+    def __init__(self, long nvars, names):
+        assert nvars >= 1
+        assert len(names) == nvars
+        self.py_names = tuple(bytes(name, 'utf-8') if not isinstance(name, bytes) else name for name in names)
+        self.c_names = <char**>libc.stdlib.malloc(nvars * sizeof(char *))
+        self._init = True
+        for i in range(nvars):
+            self.c_names[i] = self.py_names[i]
+
+    def __dealloc__(self):
+        if self._init:
+            libc.stdlib.free(self.c_names)
```

Comment:
Generally it is cleaner if we can avoid using `malloc/free` and use CPython memory management. Can we just use Python objects here rather than C-strings?

Reply:
Hmm, should we be using the `PyMem_Malloc` family of functions from the CPython API over stdlib malloc or `PyMem_RawMalloc`? I don't see any uses in the rest of the code base. This would allow our memory allocations to be on the Python heap and to utilise the `pymalloc` allocator.
https://docs.python.org/3/c-api/memory.html#c.PyMem_Malloc

With regards to using python objects, I believe this allocation is important as it sets up the C strings for later `str` calls, which may happen often, particularly in an interactive environment. Cython does more work than it appears here in that loop.
```c
    /* "flint/flint_base/flint_base.pyx":146
 *         self._init = True
 *         for i in range(nvars):
 *             self.c_names[i] = self.py_names[i]             # <<<<<<<<<<<<<<
 * 
 *     def __dealloc__(self):
 */
    __pyx_t_4 = __Pyx_GetItemInt(__pyx_v_self->py_names, __pyx_v_i, int, 1, __Pyx_PyInt_From_int, 0, 1, 1); if (unlikely(!__pyx_t_4)) __PYX_ERR(0, 146, __pyx_L1_error)
    __Pyx_GOTREF(__pyx_t_4);
    __pyx_t_8 = __Pyx_PyObject_AsWritableString(__pyx_t_4); if (unlikely((!__pyx_t_8) && PyErr_Occurred())) __PYX_ERR(0, 146, __pyx_L1_error)
    (__pyx_v_self->c_names[__pyx_v_i]) = __pyx_t_8;
    __Pyx_DECREF(__pyx_t_4); __pyx_t_4 = 0;
  }
```
I'm also not a fan of maintaining two copies of the same data, however I'm not sure of a better way here of having an easily accessible C and python version.

This section of code also generates some rather annoying warnings from flint due to non-`const` `char **`s. Not sure there is much we can do about this,
```
src/flint/types/fmpz_mpoly.c: In function ‘__pyx_pf_5flint_5types_10fmpz_mpoly_10fmpz_mpoly_4__init__’:
src/flint/types/fmpz_mpoly.c:8004:97: warning: passing argument 3 of ‘fmpz_mpoly_set_str_pretty’ from incompatible pointer type [-Wincompatible-pointer-types]
 8004 |     (void)(fmpz_mpoly_set_str_pretty(__pyx_v_self->val, __pyx_t_9, __pyx_v_self->ctx->__pyx_base.c_names, __pyx_v_self->ctx->val));
      |                                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
      |                                                                                                 |
      |                                                                                                 char **
In file included from src/flint/types/fmpz_mpoly.c:1264:
/home/jake/Uni/Honours/Thesis/local-builds/include/flint/fmpz_mpoly.h:137:49: note: expected ‘const char **’ but argument is of type ‘char **’
  137 |                                   const char ** x, const fmpz_mpoly_ctx_t ctx);
      |                                   ~~~~~~~~~~~~~~^```

Reply:
Maybe `malloc` is the right thing to use in this situation. Looking at this again I am a bit unsure about what the Cython code `self.c_names[i] = self.py_names[i]` is doing in terms of memory management. It seems like there should be a malloc there. Or is it just retrieving a pointer to the internal string data in `py_names[i]`?

What I mean by a Python object is just that Cython automatically manages ref counting for attributes of a cdef class that are Python/Cython objects (i.e. PyObject).

Maybe `c_names` should be declared as `const char ** c_names`. I always get confused about this but I think that `const char ** c_names` means that the strings are const but the pointers are not i.e. it is a non-const pointer to a non-const pointer to const string data.